src/v/rpc/parse_utils.h

```diff
@@ -109,6 +109,54 @@ struct has_rpc_adl_exempt<T, std::void_t<typename T::rpc_adl_exempt>>
 template<typename T>
 inline constexpr auto const is_rpc_adl_exempt = has_rpc_adl_exempt<T>::value;
 
+/*
+ * Encode a client request for the given transport version.
+ *
+ * Unless the message type T is explicitly exempt from adl<> support, type T
+ * must be supported by both adl<> and serde encoding frameworks. When the type
+ * is not exempt from adl<> support, serde is used when the version >= v2.
+ *
+ * The returned version indicates what level of encoding is used. This is always
+ * equal to the input version, except for serde-only messags which return v2.
+ * Callers are expected to further validate the runtime implications of this.
+ */
+template<typename T>
+ss::future<transport_version>
+encode_with_version(iobuf& out, T msg, transport_version version) {
```

Comment:
the problem with inverting it to use the case statements, i think, is that we'd couldn't constexpr exclude some case statements, and that would be necessary because some types won't have adl support and some types won't have serde support.

Reply:
but i agree with the premise that passing the version in would be easier to grok. i wonder if there is a better way...